subjects/corewar/README.md

```diff
@@ -0,0 +1,366 @@
+Player 2 ([X] bytes): [NAME] ([DESCRIPTION])
+...
+```
+
+At the end of the game the vm should print `cycle [X]: The winner is player [X]: [NAME]!`.
+
+> If nobody executed a valid `live` statement the end message should be `cycle [X]: Nobody wins!`.
+
+#### Circular memory space of the arena
+
+The memory where the players will fight is circular, this means if we want to move forward from the last address in memory (for example `4095`) we will arrive at address `0`.
+
+It implies that moving backward of one position from `0` will bring us to the address `4095`.
+
+> Having circular memory guarantees players they will never overflow the memory space they are playing in.
+
+#### Stop player execution
+
+Every `CYCLE_TO_DIE` the VM will check for every player if it signaled it was alive at least once, if it didn't all processes created by this player will be immediately killed.
+
+To avoid infinite games, `CYCLES_TO_DIE` will be decremented by `CYCLE_DELTA` under certain conditions:
+
+- If during the last life loop there were at least `NBR_LIVE` successfully executed by the players.
+- If it has been `MAX_CHECKS` life loops since it was decremented last time.
+
+> Notice that some smart players may trick another player to start making `live` statements for them, this virtually means a player can still make `live` statements after all its processes were killed.
+
+### The Assembly language
+
+In the assembly file there must be a `.name "[Name of the player]"` field and a `.description "[Description of the player]"` before any instruction.
+
+Every instruction has to be on a single line and may be preceded by a label declaration.
+The instruction name is separated from the optional label and the parameters by one or more spaces.
+
+The parameters are separated from each other by `,`.
+
+The instructions will be of different sizes depending on the number of type of parameters passed.
+
+#### Labels
+
+Labels are also called pseudo instructions, in the sense they won't be translated into binary code.
+They are widely used to help programmers creating assembly code without needing to remember and count relative memory positions.
+
+So when looking in the binaries the labels will be replaced in order to match their relative position in bytes from the place they were used.
+
+As an example in the `ameba.s` player `zjmp %:hello` is exactly the same of `zjmp %-5` (`-5` being `ff fb` in hexadecimal notation).
+
+This is because from the `zjmp` you will need to go back 5 bytes in order to match the label `hello:` declaration.
+
+> Notice that labels can be declared at the start of an instruction or on their own line.
+
+#### Parameters types
+
+- Register parameter
+
+Each process has its own registers (from 1 to `REG_NUMBER`).
+Each of those registers can hold 32 bits, but the register number itself is written in only 8 bits (which are enough to go from 1 to 16).
+
+They are formatted as `r1`, `r2`, `...`, `r16`.
```

Comment:
It's quite straightforward, but to be sure I would add that when passed as an argument, the value will be the one stored in the specified register. Additionally I would add which values are stored when the VM is initialized

Reply:
I'm not sure I understood so please correct me if I misread some part.

- The register can only pass the value that it contains... if it would pass it's register number instead it would just be a direct value (`%4` instead of `r4` for example).
On one side I understand at first glance it may not be obvious, still, adding more words to explain that may add confusion and we would be explaining why you can't do something that doesn't makes sense, which I'd rater prefer students to figure out by themselves.

- For the initialization of the register, I've added it ;)